test/functional/normal/put_spec.lua

```diff
@@ -0,0 +1,1012 @@
+                local function assertion_creator(_,_)
+                    return function(_,_)
+                        feed('u')
+                        -- Have to use feed('u') here to set curswant, because
+                        -- ex_undo() doesn't do that.
+                        eq(funcs.getcurpos(), {0, 1, 1, 0, 1})
+                        feed('<C-r>')
+                        eq(funcs.getcurpos(), {0, 1, 1, 0, 1})
+                    end
+                end
+
+                run_test_variations(
+                    create_test_defs(
+                        undo_redo_no,
+                        '<C-v>kllp',
+                        create_p_action,
+                        test_base,
+                        assertion_creator
+                    ),
+                    function() funcs.setpos('.', {0, 2, 1, 0, 1}) end
+                )
+            end)
+        end)
+
+
+        describe("with 'virtualedit'", function()
+            describe('splitting a tab character', function()
+                local base_expect_string = [[
+                Line of words 1
+                  test_stringx"     Line of words 2]]
+                run_normal_mode_tests(
+                    base_expect_string,
+                    'vp',
+                    function()
+                        funcs.setline('$', '	Line of words 2')
+                        execute('set virtualedit=all')
+                        funcs.setpos('.', {0, 2, 1, 2, 3})
+                    end,
+                    nil,
+                    ' '
+                )
+            end)
+            describe('after end of line', function()
+                local base_expect_string = [[
+                Line of words 1  test_stringx"
+                Line of words 2]]
+                run_normal_mode_tests(
+                    base_expect_string,
+                    'vp',
+                    function()
+                        execute('set virtualedit=all')
+                        funcs.setpos('.', {0, 1, 16, 2, 18})
+                    end,
+                    true,
+                    ' '
+                )
+            end)
+        end)
+    end)
+
+    describe('. register special tests', function()
+        before_each(reset)
+        it('applies control character actions', function()
+            feed('i<C-t><esc>u')
+            expect([[
+            Line of words 1
+            Line of words 2]])
+            feed('".p')
+            expect([[
+            	Line of words 1
+            Line of words 2]])
+            feed('u1go<C-v>j".p')
+            eq([[
+	ine of words 1
+	ine of words 2]], curbuf_contents())
+        end)
+
+        it('should not ring the bell with gp at end of line', function()
+            local Screen = require('test.functional.ui.screen')
+            local screen = Screen.new()
+            screen:attach()
+            helpers.ok(not screen.bell and not screen.visualbell)
+            execute('set beloff=""')
```

Comment:
Option name is misspelled. Errors in `helpers.execute()` are not bubbled up as lua errors. Each of the core functions at `test/functional/helpers.lua` have are documented to help avoid confusion about this. Basically the rule is: `feed()` (`nvim_input()`) never causes a client error, because it simulates user input. `execute()` happens to use feed(). This is admittedly confusing, I'd like to clean it up a bit...